### PR TITLE
Add Gaussian spectral model

### PR DESCRIPTION
**Pull request by JouvinLea**
Implement Gaussian model in the model Class

### CODE REVIEW

File: gammapy/spectrum/tests/test_models.py [hunk 1]
```diff
@@ -148,7 +149,21 @@ def table_model():
...
+        model=Gaussian(
+            norm=2 / u.cm ** 2 / u.s,
+            mean=4 * u.TeV,
+            sigma=3 * u.TeV,
+        ),
+
+        val_at_2TeV=u.Quantity(0.3321457437841875, 'cm-2 s-1 TeV-1'),
+        integral_1_10TeV=u.Quantity(1.6371892282407274, 'cm-2 s-1'),
+        integral_infinity=u.Quantity(2, 'cm-2 s-1'),
+        eflux_1_10TeV=u.Quantity(7.676635460998642, 'TeV cm-2 s-1'),
```
Comment (JouvinLea) [inline gammapy/spectrum/tests/test_models.py]: Maybe a better test for the energy flux?

File: gammapy/spectrum/models.py [hunk 2]
```diff
@@ -226,14 +227,14 @@ def from_dict(cls, val):
...
     def plot(
-        self,
-        energy_range,
-        ax=None,
-        energy_unit="TeV",
-        flux_unit="cm-2 s-1 TeV-1",
-        energy_power=0,
-        n_points=100,
-        **kwargs
+            self,
```
Comment (cdeil) [inline gammapy/spectrum/models.py]: There's a lot of unrelated lines that show up in this diff. Could you please run `black gammapy/spectrum/models.py gammapy/spectrum/tests/test_models.py` to go back to the black coding style?

File: gammapy/spectrum/models.py [hunk 3]
```diff
@@ -1584,3 +1584,101 @@ def evaluate(self, energy, **kwargs):
...
+
+    @staticmethod
+    def evaluate(energy, norm, mean, sigma):
+        return norm / (sigma * np.sqrt(2 * np.pi)) * np.exp((energy - mean) ** 2 / (2 * sigma ** 2))
+
+    def integral(self, emin, emax, **kwargs):
+        r"""Integrate Gaussian analytically.
+
+        .. math::
+            F(E_{min}, E_{max}) = \int_{E_{min}}^{E_{max}} \phi(E)dE = \frac{N_0}{2} \left( erf(\frac{E_{max}-
```
Comment (cdeil) [inline gammapy/spectrum/models.py]: Please break the line earlier and make it a bit easier to read.

File: gammapy/spectrum/models.py [hunk 4]
```diff
@@ -1584,3 +1584,101 @@ def evaluate(self, energy, **kwargs):
...
+            Lower and upper bound of integration range.
+        """
+        from scipy.special import erf
+
+        pars = self.parameters
+        u_min = (emin - pars['mean'].quantity) / (np.sqrt(2) * pars['sigma'].quantity)
+        u_max = (emax - pars['mean'].quantity) / (np.sqrt(2) * pars['sigma'].quantity)
+        a = pars['norm'].quantity * pars['sigma'].quantity / np.sqrt(2 * np.pi)
+        b = pars['norm'].quantity * pars['mean'].quantity / 2
+        int = a * (np.exp(- u_min ** 2) - np.exp(-u_max ** 2)) + b * (erf(u_max) - erf(u_min))
```
Comment (cdeil) [inline gammapy/spectrum/models.py]: int is a keyword, shouldn't be used a local variable name. Just change `int = ` to `return` here.

File: gammapy/spectrum/models.py [hunk 5]
```diff
@@ -1584,3 +1584,101 @@ def evaluate(self, energy, **kwargs):
...
+        int = pars['norm'].quantity / (2) * (erf(u_max) - erf(u_min))
+        return int
+
+    def energy_flux(self, emin, emax):
+        r"""Compute energy flux in given energy range analytically.
+
+        .. math::
+            G(E_{min}, E_{max}) = \int_{E_{min}}^{E_{max}}E \phi(E)dE =
+            \frac{N_0 \sigma}{\sqrt{2*\pi}}* \left( \exp(\frac{E_{min}-\bar{E}}{\sqrt{2} \sigma}) -
+            exp(\frac{E_{max} - \bar{E}}{\sqrt{2} \sigma}) \right) \\ + \frac{N_0 * \bar{E}}{2} \left( erf(\frac{E_{max}-\bar{E}}{\sqrt{2} \sigma})
```
Comment (cdeil) [inline gammapy/spectrum/models.py]: This is very hard to read. Do we need / want the equation, or could we link to Wikipedia or somewhere else instead?
If you keep it, suggest to remove `\int_{E_{min}}^{E_{max}}E \phi(E)dE =` and to break the other lines more and shorten them to make the LaTeX more readable (for review here, but also if someone looks at the help in ipython or Jupyter they will see it non-rendered)

**Comment by cdeil** [comment]
@JouvinLea - Thanks!

There's a conflict in `gammapy/spectrum/tests/test_models.py` - please resolve it by rebasing against upstream master so that this will run in CI. I also suggest to squash the commits into one or two with a good commit message. And to merge this in as soon as it's clear the equations and implementation is good.

@registerrier - Could you please review this PR and merge it in when ready?

x-ref: #1411

(unsubscribing now)

**Comment by JouvinLea** [comment]
> I put a few superficial suggestions.
> 
> I didn't check the equations and implementation yet.
> Is there a test case we could choose with a "known good answer"?
> I can't think of one, but maybe there is, e.g. on Wikipedia or somewhere?

**Comment by JouvinLea** [comment]
Hi,
Sorry for the merge conflict, I started on the morning I didn't think I could have create merge conflict in one day. Now it is fixed and I squashed also the commits!

**Comment by JouvinLea** [comment]
> I put a few superficial suggestions.
> 
> I didn't check the equations and implementation yet.
> Is there a test case we could choose with a "known good answer"?
> I can't think of one, but maybe there is, e.g. on Wikipedia or somewhere?

I don't know, I recalculate evrything my self. I think for the integral it is ok since the result is quite obvious and in the test I add a check that the integral between -inf and inf is q=equal to the norm No.

For the energy flux, it is true that the test is not a super good test since I didn't what to test to know If analytical integration was correct.

File: gammapy/spectrum/models.py (hunk 5 above)
Comment (JouvinLea) [reply gammapy/spectrum/models.py]: > This is very hard to read. Do we need / want the equation, or could we link to Wikipedia or somewhere else instead?
> If you keep it, suggest to remove `\int_{E_{min}}^{E_{max}}E \phi(E)dE =` and to break the other lines more and shorten them to make the LaTeX more readable (for review here, but also if someone looks at the help in ipython or Jupyter they will see it non-rendered)

ok, I clean a bit. I don't know on the doc I found it pretty clear to read no? I did the analytical calculation myself but I guess we could find it on wikipedia but I don't find it ugly as it is.

File: gammapy/spectrum/tests/test_models.py [hunk 6]
```diff
@@ -250,10 +258,22 @@ def test_models(spectrum):
...
+    # inverse for Gaussian has a degeneracy
+    if not (
+            isinstance(model, ConstantModel)
+            or spectrum["name"] == "compound6"
+            or spectrum["name"] == "Gaussian"
+    ):
         assert_quantity_allclose(model.inverse(value), 2 * u.TeV, rtol=0.01)
 
+    if "integral_infinity" in spectrum:
+        emin = -10000 * u.TeV
```
Comment (registerrier) [inline gammapy/spectrum/tests/test_models.py]: Shouldn't it be 0?

File: gammapy/spectrum/tests/test_models.py [hunk 7]
```diff
@@ -150,6 +151,14 @@ def table_model():
         eflux_1_10TeV=u.Quantity(6.41406327, "TeV cm-2 s-1"),
         e_peak=np.nan * u.TeV,
     ),
+    dict(
+        name="Gaussian",
+        model=Gaussian(norm=2 / u.cm ** 2 / u.s, mean=4 * u.TeV, sigma=3 * u.TeV),
```
Comment (registerrier) [inline gammapy/spectrum/tests/test_models.py]: I would also/rather test a gaussian centered at 2 TeV with a very thin width to check.

File: gammapy/spectrum/models.py [hunk 8]
```diff
@@ -1581,3 +1581,105 @@ def evaluate(self, energy, **kwargs):
...
+        pars = self.parameters
+        u_min = (emin - pars["mean"].quantity) / (np.sqrt(2) * pars["sigma"].quantity)
+        u_max = (emax - pars["mean"].quantity) / (np.sqrt(2) * pars["sigma"].quantity)
+
+        return pars["norm"].quantity / (2) * (erf(u_max) - erf(u_min))
+
+    def energy_flux(self, emin, emax):
+        r"""Compute energy flux in given energy range analytically.
+
+        .. math::
```
Comment (registerrier) [inline gammapy/spectrum/models.py]: For clarity you might give the integral in the form of 
[ primitive function ]_Emin^Emax

File: gammapy/spectrum/models.py [hunk 9]
```diff
@@ -1581,3 +1581,105 @@ def evaluate(self, energy, **kwargs):
...
+                norm
+                / (sigma * np.sqrt(2 * np.pi))
+                * np.exp((energy - mean) ** 2 / (2 * sigma ** 2))
+        )
+
+    def integral(self, emin, emax, **kwargs):
+        r"""Integrate Gaussian analytically.
+
+        .. math::
+            F(E_{min}, E_{max}) = \frac{N_0}{2} \left( erf(\frac{E_{max}-
```
Comment (registerrier) [inline gammapy/spectrum/models.py]: For clarity you could give the formula in the form of:
[primitive function]_Emin^Emax

File: gammapy/spectrum/models.py [hunk 10]
```diff
@@ -1581,3 +1581,105 @@ def evaluate(self, energy, **kwargs):
...
+
+
+class Gaussian(SpectralModel):
+    r"""Gaussian spectral model.
+
+    .. math::
+
+        \phi(E) = \frac{N_0}{\sigma \sqrt{2\pi}}  \exp{ \frac{\left( E-\bar{E} \right)^2 }{2 \sigma^2} }
+
+    This is consistent
```
Comment (registerrier) [inline gammapy/spectrum/models.py]: This is irrelevant here no?

File: gammapy/spectrum/tests/test_models.py (hunk 6 above)
Comment (JouvinLea) [reply gammapy/spectrum/tests/test_models.py]: for me the test is to check if it is equal to the norm when you integrate from -inf to +inf, no?

File: gammapy/spectrum/models.py (hunk 10 above)
Comment (JouvinLea) [reply gammapy/spectrum/models.py]: what is irrelevant?

File: gammapy/spectrum/tests/test_models.py (hunk 6 above)
Comment (JouvinLea) [reply gammapy/spectrum/tests/test_models.py]: maybe you are right and I can check that from 0 to inf is equal to No/2 as you prefer

File: gammapy/spectrum/tests/test_models.py (hunk 7 above)
Comment (JouvinLea) [reply gammapy/spectrum/tests/test_models.py]: why do you want a very thin width ?

File: gammapy/spectrum/models.py (hunk 10 above)
Comment (registerrier) [reply gammapy/spectrum/models.py]: The reference to fermipy and ctools

File: gammapy/spectrum/tests/test_models.py (hunk 6 above)
Comment (registerrier) [reply gammapy/spectrum/tests/test_models.py]: I just meant that you don't want a negative energy. So you take a small width gaussian and integration from 0 to +inf should be the same

**Comment by JouvinLea** [comment]
@registerrier 
The test are cheked with http://m.wolframalpha.com equation are ok!

**Comment by adonath** [comment]
Here are the references:
- https://www.wolframalpha.com/input/?i=integrate+1%2F+(a+*+sqrt(2+*+pi))+exp(-(x+-b)**2%2F(2*a**2))dx
- https://www.wolframalpha.com/input/?i=integrate+x%2F+(a+*+sqrt(2+*+pi))+exp(-(x+-b)**2%2F(2*a**2))dx

File: gammapy/spectrum/tests/test_models.py (hunk 7 above)
Comment (registerrier) [reply gammapy/spectrum/tests/test_models.py]: Because it is one of the test case where we might better know the expected results, and this is also the situation when the analytical integration is important

File: gammapy/spectrum/models.py [hunk 11]
```diff
@@ -1581,3 +1582,101 @@ def evaluate(self, energy, **kwargs):
...
+
+        Parameters
+        ----------
+        emin, emax : `~astropy.units.Quantity`
+            Lower and upper bound of integration range.
+        """
+        from scipy.special import erf
+
+        pars = self.parameters
+        u_min = (emin - pars["mean"].quantity) / (np.sqrt(2) * pars["sigma"].quantity)
```
Comment (registerrier) [inline gammapy/spectrum/models.py]: you should add a `.to_value('')` for `u_min` and `u_max`. 
Apparently astropy does not always handle the `erf` correctly for units.

**Comment by registerrier** [comment]
Thanks @JouvinLea .
